hw10/hw10_t01.py

```diff
@@ -0,0 +1,189 @@
+
+    Топ 10 компаний с самими дорогими акциями в рублях.
+    Топ 10 компаний с самым низким показателем P/E.
+    Топ 10 компаний, которые показали самый высокий рост за последний год
+    Топ 10 комппаний, которые принесли бы наибольшую прибыль, если бы были куплены на самом минимуме и проданы на самом максимуме за последний год.
+    Пример формата:
+
+[
+{
+    "code": "MMM",
+    "name": "3M CO.",
+    "price" | "P/E" | "growth" | "potential profit" : value,
+},
+...
+]
+"""
+import asyncio
+import datetime as dt
+import heapq
+import json
+import re
+from typing import Iterator
+
+import aiohttp
+from bs4 import BeautifulSoup
+
+NEG_INF = -float("INF")
+
+highest_price = [(NEG_INF, "", {})] * 10
+lowest_pe = [(NEG_INF, "", {})] * 10
+highest_growth = [(NEG_INF, "", {})] * 10
+highest_potential_profit = [(NEG_INF, "", {})] * 10
+
+
+async def get_sp_500_info() -> None:
+    base_url = "https://markets.businessinsider.com/index/components/s&p_500"
+    base_company_url = "https://markets.businessinsider.com"
+    async with aiohttp.ClientSession() as session:
+
+        usd = await get_usd_currency_from_cbr(session)
+
+        for page in range(1, 2):
+            async with session.get(base_url, params=[("p", page)]) as resp:
+
+                html = await resp.text()
+                soup = BeautifulSoup(html, features="html.parser")
+
+                for company in get_all_stocks_from_page(soup):
+                    name = company["title"]
+                    growth = float(
+                        list(list(company.parent.next_siblings)[-4].children)[
+                            -2
+                        ].string.rstrip("%")
+                    )
+
+                    company_url = base_company_url + company["href"]
+                    asyncio.create_task(
+                        get_company_info(company_url, usd, name, growth)
+                    )
+                    await asyncio.sleep(0.1)  #
+    write_into_json_files()
+
+
+async def get_usd_currency_from_cbr(session: aiohttp.ClientSession) -> float:
+    today_date = dt.datetime.strftime(dt.datetime.today(), "%d/%m/%Y")
+    yesterday_date = dt.datetime.strftime(
+        dt.datetime.today() - dt.timedelta(days=1), "%d/%m/%Y"
+    )
+    usd_params = {
+        "date_req1": yesterday_date,
+        "date_req2": today_date,
+        "VAL_NM_RQ": "R01235",
+    }
+    async with session.get(
+        "http://www.cbr.ru/scripts/XML_dynamic.asp", params=usd_params
+    ) as resp:
+        html = await resp.text()
+        soup = BeautifulSoup(html, features="html.parser")
+        return float(soup.find_all("value")[-1].string.replace(",", "."))
+
+
+def get_all_stocks_from_page(soup: BeautifulSoup) -> Iterator:
+    return soup.find_all("a", span="", href=re.compile("stock$"), attrs={"title": True})
+
+
+async def get_company_info(url: str, usd: float, name: str, growth: float) -> tuple:
+    async with aiohttp.ClientSession() as session:
```

Comment:
It is good practice to reuse sessions. It takes some time to establish the connection, so then you reuse it you get some performance boost. So let's try to do everything within a single session.